pinot-broker/src/main/java/org/apache/pinot/broker/routing/instanceselector/StrictReplicaGroupInstanceSelector.java

```diff
@@ -0,0 +1,179 @@
+ * (Note that the replica-group information is derived from the ideal state of the table, where the instances are sorted
+ * alphabetically in the instance state map, so the replica-groups in the instance selector might not match the
+ * replica-groups in the instance partitions.) The instances in a replica-group should have all the online segments
+ * (segments with ONLINE/CONSUMING instances in the ideal state and selected by the pre-selector) available
+ * (ONLINE/CONSUMING in the external view) in order to serve queries. If any segment is unavailable in the
+ * replica-group, we mark the whole replica-group down and not serve queries with this replica-group.
+ *
+ * The selection algorithm is the same as {@link ReplicaGroupInstanceSelector}, and will always evenly distribute the
+ * traffic to all replica-groups that have all online segments available.
+ *
+ * The algorithm relies on the mirror segment assignment from replica-group segment assignment strategy. With mirror
+ * segment assignment, any server in one replica-group will always have a corresponding server in other replica-groups
+ * that have the same segments assigned. For example, if S1 is a server in replica-group 1, and it has mirror server
+ * S2 in replica-group 2 and S3 in replica-group 3. All segments assigned to S1 will also be assigned to S2 and S3. In
+ * stable scenario (external view matches ideal state), all segments assigned to S1 will have the same enabled instances
+ * of [S1, S2, S3] sorted (in alphabetical order). If we always pick the same index of enabled instances for all
+ * segments, only one of S1, S2, S3 will be picked, and all the segments are processed by the same server. In
+ * transitioning/error scenario (external view does not match ideal state), if a segment is down on S1, we mark all
+ * segments with the same assignment ([S1, S2, S3]) down on S1 to ensure that we always route the segments to the same
+ * replica-group.
+ * </pre>
+ */
+public class StrictReplicaGroupInstanceSelector extends ReplicaGroupInstanceSelector {
+
+  public StrictReplicaGroupInstanceSelector(String tableNameWithType, BrokerMetrics brokerMetrics) {
+    super(tableNameWithType, brokerMetrics);
+  }
+
+  /**
+   * {@inheritDoc}
+   *
+   * <pre>
+   * The maps are calculated in the following steps to meet the strict replica-group guarantee:
+   *   1. Create a map from online segment to set of instances hosting the segment based on the ideal state
+   *   2. Gather the online and offline instances for each online segment from the external view
+   *   3. Compare the instances from the ideal state and the external view and gather the unavailable instances for each
+   *      set of instances
+   *   4. Exclude the unavailable instances from the online instances map
+   * </pre>
+   */
+  @Override
+  void updateSegmentMaps(ExternalView externalView, IdealState idealState, Set<String> onlineSegments,
```

Comment:
There is a need to introduce a new abstraction to simplify this class — note that the new abstraction can not be replicaGroup as it is now. We are talking about segment to replica group mapping. It should greatly simplify the codes in this class. At it current form, it looks correct to me but it is way too complex and hard to understand. We should refactor it in the next iteration.

Reply:
I think what you mean is to introduce a new algorithm for the replica-group routing. This part is quite tricky because the routing needs to still work when we change the replica-groups and run the rebalance, and the query time computation needs to be as fast as possible. If we can come up with a simpler and better algorithm, we can definitely re-implement this class.